File: Topological Sort/toposort.ml
let main () = begin
	let edges = ref [] in
	let return = ref [] in
	let no_edge = ref [] in
	let unalt_start = ref [] in
	let unalt_end = ref [] in
	

try
	while true do 
		let line = String.trim (read_line()) in
		let line2 = String.trim (read_line()) in
		unalt_end := line :: !unalt_end ;
		unalt_start := line2 :: !unalt_start ;
		edges := (line, line2) :: !edges
done

with _ -> begin
	List.iter (fun start ->
		if not (List.mem start !unalt_end) then no_edge := start :: !no_edge
	) !unalt_start ;

	while !no_edge <> [] do
		let sorted_no_edge = List.sort compare !no_edge in
		no_edge := sorted_no_edge ;
		let start = List.hd !no_edge in
		let new_no_edge = List.filter (fun x -> x <> start) !no_edge in
		no_edge := new_no_edge ;
		return := !return @ [start] ;

		List.iter (fun (a,b) ->
			if b = start then 
				let new_edges = List.filter (fun x -> x <> (a,b)) !edges in
				edges := new_edges ; 
			
				unalt_end := [] ;
				List.iter (fun (a,b) ->
					unalt_end := a :: !unalt_end;
				) !edges;

				if not (List.mem a !unalt_end) then no_edge := !no_edge @ [a] ;

		) !edges;

		
	done ;

	if (!edges <> []) then print_string "cycle"

	else List.iter (fun l -> print_string (l ^ "\n")) !return

end

end ;;
main () ;;

File: Topological Sort/toposort.py
import sys                # bring in a standard library 
lines = sys.__stdin__.readlines() # read every line from stdin into an array! 

lines = [x.strip() for x in lines]

end_nodes = []
start_nodes = []

unalt_end = []
unalt_start = []

it = iter(lines)
edges = zip(it, it)

for (a,b) in edges:
	unalt_end.append(a)
	unalt_start.append(b)

new_edges = []

for edge in edges:
	edge = list(edge)
	edge.insert(1, ord(edge[1][0]))
	edge = tuple(edge)
	new_edges.append(edge)

for edge in new_edges:
	end_nodes.append(edge[0])
	start_nodes.append(edge[2])

L = []
S = []

for node in start_nodes:
	if node not in end_nodes and node not in S:
		S.append(node)

while S != []:
	S.sort()
	n = S.pop(0)
	L.append(n)
	for edge in edges[:]:
		if edge[1] == n:
			edges.remove(edge)
			unalt_end.remove(edge[0])
			if edge[0] not in unalt_end:
				S.insert(0, edge[0])

if edges != []:
	print "cycle"

else:
	for l in L:
		print l


